aiida_quantumespresso/calculations/__init__.py

```diff
@@ -202,6 +258,35 @@ def prepare_for_submission(self, folder):
         calcinfo.uuid = str(self.uuid)
         # Empty command line by default
         cmdline_params = settings.pop('CMDLINE', [])
+
+        # Add the parallelization flags.
+        # The `cmdline_params_normalized` are used only here to check
+        # for existing parallelization flags.
+        cmdline_params_normalized = []
+        for param in cmdline_params:
+            cmdline_params_normalized.extend(param.split())
+        # To make the order of flags consistent and "nice", we use the
+        # ordering from the flag definition.
+        for flag_name in self._ENABLED_PARALLELIZATION_FLAGS:
+            # We check for existing inputs for all possible flag names,
+            # to make sure a `DeprecationWarning` is emitted whenever
+            # a flag is specified in the `settings['CMDLINE']`.
+            all_aliases = list(self._PARALLELIZATION_FLAG_ALIASES[flag_name]) + [flag_name]
+            if any(f'-{alias}' in cmdline_params_normalized for alias in all_aliases):
+                # To preserve backwards compatibility, we ignore the
+                # `parallelization.flag_name` input if the same flag is
+                # already present in `cmdline_params`. This is necessary
+                # because the new inputs can be set by their default,
+                # in user code that doesn't explicitly specify them.
+                warnings.warn(
```

Comment:
While dog-fooding this, I noticed this warning is pretty invisible. I guess it's raised somewhere in the daemon, and then promptly discarded. Maybe it'd show up if using `run` instead of `submit`.

So my question is: Is ``self.logger.warning`` the right way to warn here? Or maybe warn both via the logger and ``warnings.warn``?

In the tests, do we have a way to capture the logging and check for it?

Reply:
> While dog-fooding this

Interesting, didn't know that term yet :smile: 

> I noticed this warning is pretty invisible. I guess it's raised somewhere in the daemon, and then promptly discarded. Maybe it'd show up if using `run` instead of `submit`.

Warnings are not shown by default in Python (used to be the opposite but was changed long time ago, since it usually just annoys end-users of libraries). This is also why we have the `AiidaDeprecationWarning` in `aiida-core` to make sure that doesn't get ignored by default. You could use that as well.

> So my question is: Is ``self.logger.warning`` the right way to warn here? Or maybe warn both via the logger and ``warnings.warn``?

I wouldn't use the logger because then it will end up in the database and I don't think this deprecation warning should go on the node. We should just use the special deprecation warning class from `aiida-core` I think.

> In the tests, do we have a way to capture the logging and check for it?

I think you can use the `warning` module itself, something like:
```
with warnings.catch_warnings(record=True) as captured:
    assert 'some warning' in captured
```

Reply:
Ah, fantastic. Those end up in the daemon log I presume?

In that case the testing is also easy, we can use ``pytest.warns`` - I was just unsure about the `logging` case.

Reply:
> Ah, fantastic. Those end up in the daemon log I presume?

They are simply written to stdout, which when `run`ning is written to terminal, but for submitted jobs ends up in the daemon log yes.

> In that case the testing is also easy, we can use `pytest.warns`

Yes, this is also possible since it is just using Python's logging system

Reply:
> Yes, this is also possible since it is just using Python's logging system

I think ``warnings.warn`` and ``logging`` are two different things, no? It seems for `logging` the `caplog` fixture needs to be used: https://docs.pytest.org/en/6.1.1/logging.html

Anyway, the code is updated to use `AiidaDeprecationWarning` now.

Reply:
Yeah, I meant the `warnings` system of course.